Allows user to configure the maximum fee

robbiehanson opened the pull request:
Fixes issue #191

This PR requires [13c7143](https://github.com/ACINQ/lightning-kmp/commit/13c71436daa1fc8f72166244c4aa143e958c3544) from https://github.com/ACINQ/lightning-kmp/pull/317

<img height="450" alt="Screen Shot 2022-01-24 at 11 49 50" src="https://user-images.githubusercontent.com/304604/150827762-3ffdcbed-3640-4511-b05e-dfb2c561f54c.png">

## Technical notes:

My theory is that users will think of this feature a bit like an Ebay auction. With Ebay, if I make a bid of X, I know that I'll only pay Y, where Y is the minimum necessary amount to win the bid. Similarly, if a user sets the max fee to X, they're expecting to pay Y, where Y is close to the minimum fee necessary to make the payment succeed.

So the lightning-kmp library needs a schedule of fees. And our default schedule is a slow ramp up to a conservative max fee. My fear is that, if we simply append the max fee of X at the end of the list, then the user will be paying the max fee everytime the required fee exceeds the conservative maximum.

So I'm thinking that, if our default schedule is something like `[Q, R, S, T, U]`, and the user sets the max to X, then the ideal schedule is more like `[Q, R, S, T, U, V, W, X]`, where `[V, W]` ramp up to X. So there's a bit of clever logic in `TrampolineFees.kt` to accomplish this, along with associated unit tests.

dpad85 commented on phoenix-shared/src/commonTest/kotlin/fr/acinq/phoenix/utils/TrampolineFeesTests.kt:
```diff
@@ -0,0 +1,102 @@
+                TrampolineFees(65.sat, 1800, CltvExpiryDelta(576)),
+                TrampolineFees(95.sat, 2100, CltvExpiryDelta(576))
+            ))
+        }
```
We could also check behaviour when the proposed base fee is over the significant diff threshold, but proportional fee is below the default max fee (so default list is truncated). And vice-versa.

dpad85 reviewed (commented):
Looks good. I would add a sanity check somewhere. Proportional fee should not be exceed 100% for example. When that happens, an error is shown.

robbiehanson commented:
Good idea. I'm also thinking an example might be helpful. (Math is hard.) Kinda like we did with the tip summary on the outgoing payment screen.

robbiehanson commented:
UI has been improved. First, I added a quick explanation:

<img height="450" alt="explanation" src="https://user-images.githubusercontent.com/304604/151587613-f0414e65-97dc-4f75-a992-b9a12e9fa721.png">

Then added an example. Makes the math & fees easier to grok. Note that the example is interactive - user can use plus/minus buttons to see how fees affect payments of different sizes.

<img height="450" alt="example" src="https://user-images.githubusercontent.com/304604/151587811-3ba7cce4-3c88-4897-b294-362c32ece029.png">

Added sanity checks to ensure baseFee & proportionalFee are within min/max limits. (This is configurable. I'm not really sure what exact values we want to enforce here. Let me know what you think.)

<img height="450" alt="aboveMax" src="https://user-images.githubusercontent.com/304604/151588285-8fa7ba3a-b9f3-41b0-82e5-b20e2a1ffbb1.png">

And added checks as recommended concerning odd mismatches:

<img height="450" alt="badBase" src="https://user-images.githubusercontent.com/304604/151589007-40f9f53b-ef76-4fac-b7f3-de9359a49c07.png">

dpad85 reviewed (approved):
LGTM. The example section is nice 👍
